DNS configuration guidelines (#239)
pyansys/dev-guide: doc/source/how-to/dns-configuration.rst

DNS configuration
=================

As explained in :ref:`Documenting`, PyAnsys projects publish their documentation
online under the following canonical name (CNAME) convention:

``https://<product>.docs.pyansys.com``

In order to request a CNAME for the ``pyansys.com`` domain, please contact
`Maxime Rey`_, `Roberto Pastor Muela`_ or `Alex Kaszynski`_. Any of these members
can handle the creation of the requested PyAnsys subdomain.

Once the CNAME is created, repository administrators can configure their published

> documentation in GitHub pages to be exposed through it. In order to configure the CNAME

Reviewer: Maybe here we could specify that in one hand the CNAME is located the `gh-pages` branch.
But on the other hand the Custom domain is also mentioned in GitHub repository `Settings/Pages `.

Talking about those two distinct locations where the custom domain is mentioned might help to clarify the process because I noticed some confusion for internal users.

Maybe not the perfect location to talk about it but it might be helpful.

Author: CNAME configuration is done through Settings/Pages --> Custom domain.

Once you do that, GitHub automagically creates the ``CNAME`` file in your ``gh-pages`` branch without you having to do anything... Or in whatever branch your docs are published. Remember that the naming convention ``gh-pages`` is only that, a convention. You could potentially publish your docs under a branch called ``my-docs-branch``, for example.

I don't know if it's worth mentioning that the CNAME is also located in the ``gh-pages`` branch or not honestly.. because admins won't have to do anything there. It's done for them when you configure it in the Settings/Pages site. And if you change it, it also gets updated in the ``gh-pages`` branch.

But I'm open to discussion honestly 😄

Reviewer: You are correct, thinking twice about it, it could bring more confusion.

Reviewer 2: Also, you can publish now with a CI action without having a branch at all.

Author: Yep, that's also true. GH evolves fast 😄

DNS configuration
=================

As explained in :ref:`Documenting`, PyAnsys projects publish their documentation
online under the following canonical name (CNAME) convention:

``https://<product>.docs.pyansys.com``

> To request a CNAME for the ``pyansys.com`` domain, please contact

Reviewer 3: ```suggestion
To request a CNAME for the ``pyansys.com`` domain, contact
```
Avoid the use of "please" in technical documentation--according to all style guides!

Author: Thanks Kathy!!

> `Maxime Rey`_, `Roberto Pastor Muela`_ or `Alex Kaszynski`_. Any of these members

Reviewer 4: ```suggestion
`Maxime Rey_`_, `Roberto Pastor Muela_`_ or `Alex Kaszynski_`_. Any of these members
```

> `Maxime Rey`_, `Roberto Pastor Muela`_ or `Alex Kaszynski`_. Any of these members

Reviewer 3: Given that this guide is public-facing, do we want to use specific names or use the PyAnsys support email address?

Author: This specific part of the documentation, even if it's public facing, is directed for Ansys employees only, since they are the ones that can request CNAME creations. I think it's fine if we leave it as is.

can handle the creation of the requested PyAnsys subdomain.

Once the CNAME is created, repository administrators can configure their published
documentation in GitHub pages to be exposed through it. To configure the CNAME

> for your documentation, please refer to GitHub's online documentation:
> `Managing a custom domain for your GitHub Pages site`_.

Reviewer 3: ```suggestion
for your documentation, refer to `Managing a custom domain for your GitHub Pages site`_.
```

PyAnsys verified domains
------------------------

> In the `PyAnsys GitHub organization`_, the following domains have been verified:

Reviewer 3: ```suggestion
In the `PyAnsys GitHub organization`_, these domains have been verified:
```

* ``pyansys.com``
* ``docs.pyansys.com``

.. warning::

    Only CNAME requests with **one** subdomain before the previous verified
    domains are allowed. The reasons behind this measure are explained in
    :ref:`DNS protection measures`.

DNS protection measures
-----------------------

The rationale behind choosing the previous CNAME convention is due to cybersecurity reasons.

> GitHub allows to verify domains for users and organizations, see `Verifying a domain for your organization site`_.

Reviewer 3: ```suggestion
As explained in `Verifying a domain for your organization site`_, GitHub provides for
verifying domains for users and organizations.
```

Once an organization has a verified domain, **this prevents external users to the organization from
taking over existing direct subdomains**. However, deeper subdomains are no longer verified by
GitHub.

This is better explained with the following examples:

Case scenario - **protected** subdomain
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> 

Reviewer 3: I'm changing the lists to bullets as these aren't procedural (ordered) steps.

Author: Sure, thanks @PipKat!!

> #. Consider that the domain ``pyansys.com`` has been verified for the `PyAnsys GitHub organization`_.
> #. A CNAME as follows is requested: ``subdomain.pyansys.com``

Reviewer 3: ```suggestion
- Consider that the domain ``pyansys.com`` has been verified for the `PyAnsys GitHub organization`_.
- This CNAME is requested: ``subdomain.pyansys.com``.
```

> The CNAME requested can only be used by projects/repositories inside the `PyAnsys GitHub organization`_.
> Any attempt by external users/organizations of taking over the previous CNAME is identified and rejected by GitHub.

Reviewer 3: ```suggestion
This CNAME can only be used by repositories inside the `PyAnsys GitHub organization`_.
Any attempt by an external user to take over this CNAME is identified and rejected by GitHub.
```
Avoid slashes when possible. Pick the most descriptive choice.

Case scenario - **vulnerable** subdomain
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> #. Consider that the domain ``pyansys.com`` has been verified for the `PyAnsys GitHub organization`_.
> #. A CNAME as follows is requested: ``subsubdomain.subdomain.pyansys.com``
> 
> The CNAME requested **can** be used by external user/organization repositories. For this reason,
> it must be avoided to create CNAME requests that are not verified by the organization.

Reviewer 3: ```suggestion
- The domain ``pyansys.com`` has been verified for the `PyAnsys GitHub organization`_.
- This CNAME is requested: ``subsubdomain.subdomain.pyansys.com``.

This CNAME **can** be used by external users for their repositories. For this reason,
you must avoid creating CNAME requests that are not verified by the organization.
```

Preventing CNAME takeover
-------------------------

> From past experience, CNAME values have been taken over by external users/organizations. This has typically
> occurred due to the following reasons:

Reviewer 3: ```suggestion
CNAME values have been taken over in the past by external users, typically due to these reasons:
```

* Ansys GitHub organizations had no domain verification set up.
* A CNAME created did not follow the recommended CNAME guidelines.
* More than one level of subdomain depth under the verified domain had been requested.

> * Long lead times between CNAME creation and assignment to GitHub pages occurred.

Reviewer 3: ```suggestion
* Long time lapses occurred between CNAME creation and assignment to GitHub pages.
```

> Thus, it is important that the following guidelines are followed:

Reviewer 3: ```suggestion
Thus, it is important that you follow these guidelines:
```

* Ensure that your GitHub organization has verified domains for hosting GitHub pages.

> * Check that whatever CNAME requested does not have a subdomain depth larger than **1** with respect to the verified domains.

Reviewer 3: ```suggestion
* Check that the CNAME that you request does not have a subdomain depth larger than **1** with respect to the verified domains.
```

> * Request a CNAME only when needed - that is, right before publishing the site.

Reviewer 3: ```suggestion
* Request a CNAME only when needed, which is just prior to publishing the site.
```

> * Request CNAME deletion once it is no longer used to prevent others from hosting their sites on it.

Reviewer 3: ```suggestion
* Request deletion of the CNAME once it is no longer used to prevent others from hosting their sites on it.
```

DNS configuration
=================

As explained in :ref:`Documenting`, PyAnsys projects publish their documentation
online under the following canonical name (CNAME) convention:

``https://<product>.docs.pyansys.com``

> In order to request a CNAME for the ``pyansys.com`` domain, please contact

Reviewer 5: ```suggestion
To request a CNAME for the ``pyansys.com`` domain, please contact
```

Reviewer 5: I tend to use "in order to..." too. However, as per @PipKat reviews, I know most of the times its usage can be avoided as it does not introduce any information.

`Maxime Rey`_, `Roberto Pastor Muela`_ or `Alex Kaszynski`_. Any of these members
can handle the creation of the requested PyAnsys subdomain.

Once the CNAME is created, repository administrators can configure their published

> documentation in GitHub pages to be exposed through it. In order to configure the CNAME

Reviewer 5: ```suggestion
documentation in GitHub pages to be exposed through it. To configure the CNAME
```

for your documentation, please refer to GitHub's online documentation:
`Managing a custom domain for your GitHub Pages site`_.

PyAnsys verified domains
------------------------

> In the `PyAnsys organization`_, the following domains have been verified:

Reviewer 5: Suggesting to use `PyAnsys GitHub organization` here, as this term appears in other parts of the guide:

```suggestion
In the `PyAnsys GitHub organization`_, the following domains have been verified:
```

* ``pyansys.com``
* ``docs.pyansys.com``

.. warning::

    Only CNAME requests with **one** subdomain before the previous verified
    domains are allowed. The reasons behind this measure are explained in
    :ref:`DNS protection measures`.

DNS protection measures
-----------------------

The rationale behind choosing the previous CNAME convention is due to cybersecurity reasons.
GitHub allows to verify domains for users and organizations. In order to do so, GitHub's online
documentation can be found here: `Verifying a domain for your organization site`_.

Once an organization has a verified domain, **this prevents external users to the organization from
taking over existing direct subdomains**. However, deeper subdomains are no longer verified by
GitHub.

This is better explained with the following examples:

Case scenario - **protected** subdomain
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#. Consider that the domain ``pyansys.com`` has been verified for the `PyAnsys organization`_.
#. A CNAME as follows is requested: ``subdomain.pyansys.com``

The CNAME requested can only be used by projects/repositories inside the `PyAnsys organization`_.

> Any attempt by external users/organizations of taking over the previous CNAME is taken
> care by GitHub and rejected.

Reviewer 5: WHat about rewriting this as to avoid using "take" twice? 

```suggestion
Any attempt by external users/organizations of taking over the previous CNAME is identifiead and rejected by GitHub.
```
